### PR TITLE
[TOP][ARM] Winograd nnpack backend

### PR DESCRIPTION
**Pull request by hlu1**
Discussions #2692
@merrymercy, @ajtulloch, @yidawang please review.

### CODE REVIEW

File: topi/python/topi/arm_cpu/conv2d.py [hunk 1]
```diff
@@ -499,6 +508,70 @@ def _schedule_winograd(cfg, s, output, last):
...
+    with tvm.tag_scope("winograd_nnpack_conv2d_output"):
+        output = tvm.contrib.nnpack.convolution_inference_without_weight_transform(
+            data, transformed_kernel,
+            bias=None,
+            padding=[HPAD, HPAD, WPAD, WPAD],
+            stride=[HSTR, WSTR],
+            algorithm=cfg['winograd_nnpack_algorithm'].val)
+
+    # we have to manually assign effective GFLOP for winograd
+    cfg.add_flop(2 * N * CI * H * W * KH * KW * CO)
```
Comment (yidawang) [inline topi/python/topi/arm_cpu/conv2d.py]: Out of curiosity, is it because nnpack needs this info to decide if winograd should kick in?

Comment (hlu1) [reply topi/python/topi/arm_cpu/conv2d.py]: There are two reasons:
1) NNPACK is an extern op, `compute_flops` does not know how to compute the flops for extern ops
2) The winograd implementation reduces the algorithmic complexity by reducing the number of multiplication operators, so here we care about the "fake GFlops" (the number of multiplys as if we did a direct conv) so later AutoTVM can choose the best out of direct, winograd, winograd_nnpack.

File: topi/python/topi/arm_cpu/conv2d.py [hunk 2]
```diff
@@ -522,6 +595,58 @@ def _callback(op):
     return s
 
 
+##### REGISTER TOPI COMPUTE / SCHEDULE FOR WINOGRAD NNPACK WITH WEIGHT TRANSFORM #####
```
[the remark this answers is not in the document]
Comment (hlu1) [reply topi/python/topi/arm_cpu/conv2d.py]: Nice catch

**Comment by hlu1** [comment]
To answer your question:
1) Yes. And I would like to get some early feedbacks on the topi/nnpack implementation first.
2) We need to separate the weight transform out into another node so that it can be precomputed as part of the pregraph (the PrecomputePrune pass in nnvm). This technique is used by the TVM winograd implementation as well (https://github.com/dmlc/tvm/blob/master/topi/python/topi/arm_cpu/conv2d.py#L503).

**Comment by hlu1** [comment]
@tqchen and @merrymercy, I noticed that in relay, bias is no longer part of the conv op. However, in NNPACK bias is a required input for conv. What's the easiest way to map conv + bias_add in relay to conv in NNPACK? The most obvious way is probably to pass zero_bias to nnpack (less preferred because it incurs perf overhead), I'm wondering if there's a cleaner way to do this.

**Comment by FrozenGene** [comment]
Maybe you have to pass zero_bias to it. I can not think of simple and clean way too, because they are two isolated ops. I am also interested in knowing some better way too.

**Comment by tqchen** [comment]
@hlu1 we could introduce another experimental op that contains bias, and mark it as internal (level=10), then allow pattern matching to the new one. Alternatively, we can start with zero bias

**Comment by hlu1** [comment]
@tqchen, can you elaborate a bit more on the first approach? What do you mean by allowing pattern matching to the new one?

**Comment by tqchen** [comment]
Yes, what I mean is to transform conv->bias_add => ConvWithBias

**Comment by hlu1** [comment]
I fixed the bias in relay by passing `None` to `topi.nn.conv2d_winograd_nnpack_without_weight_transform` ( https://github.com/dmlc/tvm/pull/2721/files#diff-b8a5824a5a8f1907448d2588d8cd773cR326)

File: include/tvm/relay/attrs/nn.h [hunk 3]
```diff
@@ -155,6 +155,23 @@ struct Conv2DWinogradAttrs : public tvm::AttrsNode<Conv2DWinogradAttrs> {
...
+struct Conv2DWinogradNNPACKWeightTransformAttrs
+    : public tvm::AttrsNode<Conv2DWinogradNNPACKWeightTransformAttrs> {
+  int convolution_algorithm;
+  DataType out_dtype;
+
+  TVM_DECLARE_ATTRS(Conv2DWinogradNNPACKWeightTransformAttrs,
+                    "relay.attrs.Conv2DWinogradNNPACKWeightTransformAttrs") {
+    TVM_ATTR_FIELD(convolution_algorithm)
+        .describe(
+            "The convolution algorithm for Winograd NNPACK. E.g. 3 for WT_8x8, "
```
Comment (ajtulloch) [inline include/tvm/relay/attrs/nn.h]: Could you explicitly reference the enum defined in `tvm.contrib.nnpack.ConvolutionAlgorithm`?

File: nnvm/include/nnvm/top/nn.h [hunk 4]
```diff
@@ -183,6 +183,25 @@ struct WinogradWeightTransformParam : public dmlc::Parameter<WinogradWeightTrans
...
 
+struct WinogradNNPACKWeightTransformParam
+    : public dmlc::Parameter<WinogradNNPACKWeightTransformParam> {
+  int convolution_algorithm;
+  int out_dtype;
+
+  DMLC_DECLARE_PARAMETER(WinogradNNPACKWeightTransformParam) {
+    DMLC_DECLARE_FIELD(convolution_algorithm)
+        .describe(
+            "The convolution algorithm for Winograd NNPACK. E.g. 3 for WT_8x8, "
```
Comment (ajtulloch) [inline nnvm/include/nnvm/top/nn.h]: Ditto re: `tvm.contrib.nnpack.ConvolutionAlgorithm`.

File: src/relay/op/nn/convolution.cc [hunk 5]
```diff
@@ -10,6 +10,7 @@
 #include "../../pass/alter_op_layout.h"
 #include "../layout.h"
 
+
```
Comment (ajtulloch) [inline src/relay/op/nn/convolution.cc]: Remove?

File: src/relay/op/nn/convolution.cc [hunk 6]
```diff
@@ -517,6 +519,123 @@ weight transformation in advance.
...
+- **data**: Input is 4D array of shape  (batch_size, in_channels, height, width)
+- **weight**: Any shape
+            We do not check the shape for this input tensor. Since different backend
+            has different layout strategy.
+
+- **out**:  Output is 4D array of shape (batch_size, channels, out_height, out_width)
+)code" TVM_ADD_FILELINE)
+.set_attrs_type_key("relay.attrs.Conv2DAttrs")
+.set_num_inputs(2)
```
Comment (ajtulloch) [inline src/relay/op/nn/convolution.cc]: Curious, why doesn't this also have an optional bias input?

File: topi/python/topi/arm_cpu/injective.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
+# pylint: disable=invalid-name, unused-variable
```
Comment (ajtulloch) [inline topi/python/topi/arm_cpu/injective.py]: Were these intended to be in this diff? They seem useful but should be in a separate PR right?

File: topi/python/topi/arm_cpu/pooling.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
+# pylint: disable=invalid-name, unused-variable
+"""Schedule for pooling operators"""
```
Comment (ajtulloch) [inline topi/python/topi/arm_cpu/pooling.py]: Ditto re: a separate PR?

File: include/tvm/relay/attrs/nn.h (hunk 3 above)
Comment (hlu1) [reply include/tvm/relay/attrs/nn.h]: Yes, will do

File: nnvm/include/nnvm/top/nn.h (hunk 4 above)
Comment (hlu1) [reply nnvm/include/nnvm/top/nn.h]: Yep

File: src/relay/op/nn/convolution.cc (hunk 5 above)
Comment (hlu1) [reply src/relay/op/nn/convolution.cc]: Will do

File: src/relay/op/nn/convolution.cc (hunk 6 above)
Comment (hlu1) [reply src/relay/op/nn/convolution.cc]: Because the conv2d op in relay does not have bias.

File: topi/python/topi/arm_cpu/injective.py (hunk 7 above)
Comment (hlu1) [reply topi/python/topi/arm_cpu/injective.py]: Will do

File: topi/python/topi/arm_cpu/pooling.py (hunk 8 above)
Comment (hlu1) [reply topi/python/topi/arm_cpu/pooling.py]: Will do

**Comment by hlu1** [comment]
@ajtulloch, I removed `injective.py` and  `pooling.py` from this PR and added unet tests for both nnvm and relay

**Comment by hlu1** [comment]
@tqchen, is `tests/python/integration/test_winograd_nnpack.py` a good place for this test? It's not exactly for testing frontend, that's why I didn't put it under `tests/python/frontend/mxnet/`

**Comment by hlu1** [comment]
Separate the unet e2e test out as a separate PR. The topi unit test is included though.

**Comment by hlu1** [comment]
@tqchen, the test I added `topi/tests/python/test_topi_conv2d_winograd_nnpack.py` seems to be failing because of this error:
```
Cannot find function tvm.contrib.nnpack.convolution_inference_weight_transform in the imported modules or global registry
```
But NNPACK is enabled in the build. Do you know how I can fix it?

**Comment by tqchen** [comment]
Please confirm that the cmake rule actually includes the files that register the function.

**Comment by tqchen** [comment]
The docker is maintained separated, see https://docs.tvm.ai/contribute/pull_request.html#ci-environment

Please also confirm if you need to run the test in GPU env, or if CPU is fine and you can safely skip it in GPU CI

**Comment by hlu1** [comment]
@tqchen, the topi tests run in the ci-gpu.
``` 
[topi-python-gpu] Running shell script
+ docker/bash.sh tvmai/ci-gpu:v0.50 ./tests/scripts/task_python_topi.sh
```

**Comment by tqchen** [comment]
@yidawang @ajtulloch pleas another look and https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly

File: python/tvm/relay/op/nn/nn.py [hunk 9]
```diff
@@ -1013,3 +1080,28 @@ def contrib_conv2d_winograd_weight_transform(weight,
...
+
+    We separate this as a single op to enable pre-compute for inference.
+    Use this together with nn.contrib_conv2d_winograd_without_weight_transform
+
+    Parameters
+    ----------
+    weight : tvm.relay.Expr
+        The weight expressions.
+
+    tile_size : int
```
Comment (ajtulloch) [inline python/tvm/relay/op/nn/nn.py]: Needs to be updated with an explanation of the `convolution_algorithm` parameter right?

File: src/relay/op/nn/convolution.cc [hunk 10]
```diff
@@ -530,6 +531,123 @@ weight transformation in advance.
...
+  const auto* data = types[0].as<TensorTypeNode>();
+  if (data == nullptr) return false;
+
+  const Conv2DWinogradNNPACKWeightTransformAttrs* param =
+      attrs.as<Conv2DWinogradNNPACKWeightTransformAttrs>();
+  CHECK(param != nullptr);
+
+  CHECK_EQ(data->shape.size(), 4) << "Only support NCHW normal kernel layout";
+
+  // each pad width element should be a pair of positive integers
```
Comment (ajtulloch) [inline src/relay/op/nn/convolution.cc]: What does this mean?

File: python/tvm/relay/op/nn/nn.py (hunk 9 above)
Comment (hlu1) [reply python/tvm/relay/op/nn/nn.py]: Nice catch.

File: src/relay/op/nn/convolution.cc (hunk 10 above)
Comment (hlu1) [reply src/relay/op/nn/convolution.cc]: It was copied from Conv2DWinogradWeightTransformRel. Now removed.

**Comment by hlu1** [comment]
@tqchen, all comments have been addressed. It should be good to go.
